qasper_baselines/model.py

```diff
@@ -154,15 +160,60 @@ def forward(
                 device=evidence_logits.device,
                 dtype=evidence_logits.dtype,
             )
-            loss_fn = torch.nn.CrossEntropyLoss(weight=weights)
-            evidence_loss = loss_fn(evidence_logits.view(-1, 2), evidence.view(-1))
-            self._evidence_loss(float(evidence_loss.detach().cpu()))
+            if self._use_margin_loss_for_evidence:
+                evidence_probs = torch.softmax(evidence_logits, dim=1)
+
+                positive_example_indices = (evidence == 1).nonzero()[:, 1].unsqueeze(0)
+                positive_example_scores = util.batched_index_select(evidence_probs, positive_example_indices).squeeze(0)
+                min_positive_score = torch.min(positive_example_scores).unsqueeze(0)
+
+                negative_example_indices = (evidence == 0).nonzero()[:, 1].unsqueeze(0)
+                negative_example_scores = util.batched_index_select(evidence_probs, negative_example_indices).squeeze(0)
+                max_negative_score = torch.max(negative_example_scores).unsqueeze(0)
+
+                null_score = evidence_probs[:, 0, :].reshape(max_negative_score.size())
+
+                if self.training:
+                    min_positive_index = positive_example_indices[:, torch.argmin(positive_example_scores)]
+                    min_positive_prob = evidence_probs[:, min_positive_index, :]
+                    max_negative_index = negative_example_indices[:, torch.argmax(negative_example_scores)]
+                    max_negative_prob = evidence_probs[:, max_negative_index, :]
+                    if max_negative_prob < min_positive_prob:
+                        self._train_evidence_logit_thresh(max_negative_prob+(min_positive_prob-max_negative_prob)/2)
+                    else:
+                        self._train_evidence_logit_thresh(min_positive_prob+(max_negative_prob-min_positive_prob)/2)
```

Comment:
This is the case where some positives are scored lower than some negatives, right? I am not sure if this should be the threshold then. Why not make the threshold always be `min(min_positive_prob, max_negative_prob)`?